File: P20 Valid Parentheses/Solution.java
class Solution {

    private static final Map<Character, Character> bracketMap = Map.of('(', ')', '[', ']', '{', '}');

    // Time: O(N)
    // Space: O(N)
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        for(char c: s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else if (stack.empty() || c == ')' && stack.pop() != '(' || c == '}' && stack.pop() != '{' || c == ']' && stack.pop() != '[') {
               return false;
            }
        }
        return stack.empty();  
    }

    // Time: O(N)
    // Space: O(N)
    public boolean isValid2(String s) {
        if (s.length() % 2 != 0) return false;
        
        Stack<Character> endBracketStack = new Stack<>();
        
        for (char c : s.toCharArray()) {
            if (bracketMap.containsKey(c)) {
                endBracketStack.addLast(bracketMap.get(c));
            } else {
                if (endBracketStack.isEmpty() || endBracketStack.removeLast() != c) {
                    return false;
                }
            }
        }
        return endBracketStack.isEmpty();
    }
}

File: P20 Valid Parentheses/solution.py
class Solution:
    # Time : O(N)
    # Space : O(N)    
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.append(c)
            elif len(stack) == 0 or c == ')' and stack.pop() != '(' or c == '}' and stack.pop() != '{' or c == ']' and stack.pop() != '[':
                return False
        return len(stack) == 0
